Avoid calculating dynamic lights when lights are already baked using the static bake mode in the Forward+ renderer

clayjohn opened the pull request:
Fixes an issue related to: https://github.com/godotengine/godot/issues/85145

When baking a light with the static bake mode, the light is calculated twice. Once when baking and once at runtime. The light values are added together resulting in double lighting. 

Shadows are already skipped using the exact same code, but the actual lighting was still being added.

Opening as draft as I need to check the mobile renderer and the compatibility renderer

clayjohn committed:
Avoid calculating dynamic lights when lights are baked into LightmapGI using the static bake mode

Calinou commented:
There's already a skip in place here for directional lights: https://github.com/godotengine/godot/blob/694d3c2930bdfb43fd93e1e6641f66c4f19a5c77/servers/rendering/renderer_rd/shaders/forward_clustered/scene_forward_clustered.glsl#L1853-L1855

The same skip also already exists for OmniLight3D and SpotLight3D (search for the comment string).

clayjohn commented:
That other check only disables shadows. The lighting was added in either case.

For the mobile renderer the check shouldn't be in the shader, it should be excluding the light from the light list when pairing with the object so we don't get any overhead from the light at all

clayjohn commented:
@Calinou Updated!

In the end this was the situation:
1. Forward+ was missing the check for the light portion (it was only checked for shadows and the flag was only passed in when shadows were enabled)
2. Mobile was missing the check everywhere. The correct space to add it is in the shader for directional lights, but during pairing for positional lights
3. Compatibility did the check for positional lights after pairing, so the check became redundant. 

It should be ready to go now!

Calinou reviewed (commented):
Tested locally, it mostly works as expected in all rendering methods. Performance increases as expected when baking with the Static bake mode as opposed to the default Dynamic.

**Testing project:** [test_lightmapgi_mobile.zip](https://github.com/godotengine/godot/files/13417782/test_lightmapgi_mobile.zip)

There's one issue where DirectionalLight3D with shadow disabled still applies double lighting, but only in Compatibility:

### DirectionalLight3D shadow enabled (Compatibility)

![image](https://github.com/user-attachments/assets/e754f80c-e7b1-4f41-bc79-828dacd0824a)

### DirectionalLight3D shadow enabled (Compatibility)

![image](https://github.com/user-attachments/assets/06f5083a-148c-48f0-b779-7e2f1c47fd90)

OmniLight3D and SpotLight3D don't have this issue in Compatibility.

clayjohn commented:
@Calinou Thanks for checking! It turned out that the vertex lighting codepath in the GLES3 shader was missing the check. I added it and removed the static checks for omnilights and spotlights since they weren't necessary anymore. 

This should be ready to go now!

Calinou reviewed (approved):
Works great now :slightly_smiling_face:

AThousandShips commented on servers/rendering/renderer_rd/shaders/forward_clustered/scene_forward_clustered.glsl:
```diff
@@ -2211,6 +2211,10 @@ void fragment_shader(in SceneData scene_data) {
 			}
 
+			if (directional_lights.data[i].bake_mode == LIGHT_BAKE_STATIC && bool(instances.data[instance_index].flags & INSTANCE_FLAGS_USE_LIGHTMAP)) {
+				continue; // Statically baked light and object uses lightmap, skip
```
```suggestion
				continue; // Statically baked light and object uses lightmap, skip.
```

Repiteo commented:
Thanks!
